RuntimeTypeInspector.js: Add RTI build target + Examples integration

kungfooman opened the pull request:
Fixes #2529

I'm excited to introduce [RuntimeTypeInspector.js](https://runtimetypeinspector.org/), which I believe will benefit PlayCanvas as a whole and our debugging workflows in specific. RTI is designed to type-check the arguments of every function and method. It detects type errors while the engine is running in a way that goes beyond the capabilities of TypeScript, as demonstrated in [this video](https://www.youtube.com/watch?v=o5ipQe2rVKQ) (there is another one on the website debugging the Microsoft Holometric Video PC project).

During the development of RTI it already helped me to open issues or make PR's for various issues (e.g. NaN bugs).

I believe that RTI will help improve productivity and efficiency for both engine developers and users, by catching errors early on and reporting them in detail.

Runtime Type Inspector is implemented using Babel AST traversal with custom JSDoc parsing (powered by TypeScript itself for highest compatibility), customized to PlayCanvas requirements. That means for example supporting polymorphing `this` types.

TLDR: It adds type assertions for runtime type checking, for example:

![image](https://github.com/playcanvas/engine/assets/5236548/e4363d79-a34f-4d38-8ea9-1a73b9536f7e)

Being able to set breakpoints exactly where the errors are happening for the first time helps to track down problems that are otherwise difficult to pinpoint (and saves a lot of time :sweat_smile:).

Try for yourself: http://pc.runtimetypeinspector.org/

Just make sure to pick the right JS context - the exampleIframe:

![image](https://github.com/playcanvas/engine/assets/5236548/df2d8a9a-4b88-4f67-9942-3ad0b7a63632)

I'm working on this for quite some time by now, a mix of fun and frustration - rewriting code on the AST level first seemed tricky, but given enough time, it was a rather straightforward way to automate type validations. I'm excited to read feedback and suggestions. :pray: 

I confirm I have read the [contributing guidelines](https://github.com/playcanvas/engine/blob/master/.github/CONTRIBUTING.md) and signed the [Contributor License Agreement](https://docs.google.com/a/playcanvas.com/forms/d/1Ih69zQfJG-QDLIEpHr6CsaAs6fPORNOVnMv5nuo0cjk/viewform).

kungfooman committed:
Adding RTI transpiler to every target, I will fine-tune a bit later

kungfooman committed:
Merge branch 'main' into rti

kungfooman committed:
name output build/playcanvas-rti and update packages

kungfooman committed:
Merge branch 'main' into rti

kungfooman committed:
Resolve RTI runtime and update to latest RTI rollup plugin

kungfooman committed:
Integrate RTI into Examples browser via `npm run develop:rti`

kungfooman committed:
only add node-resolve plugin for RTI

kungfooman committed:
Don't type check framework/parsers/draco-worker.js (because a Worker context has no access to RTI)

kungfooman committed:
npm i @runtime-type-inspector/plugin-rollup@1.0.6 (added `types.d.ts` for IntelliSense)

Maksims commented:
This is pretty cool!
A couple of questions:
1. Is there a way to suppress reporting more than N errors if they are originating from the same place? Spam of errors on every application update - is not much of use.
2. Is there a way to improve the stack within console log reporting so it points to the right place of the error? This will allow easier debugging with Dev Tools, but also provide a deep linking from Launcher to IDE.

kungfooman committed:
Merge branch 'main' into rti

kungfooman commented:
Hey @Maksims, thank you for the feedback!

> 1. Is there a way to suppress reporting more than N errors if they are originating from the same place? Spam of errors on every application update - is not much of use.

Yep, for this there is this button:

![image](https://github.com/playcanvas/engine/assets/5236548/1bdc68ac-f880-455d-af9b-fd3cfe508159)

Currently it supports a "warn once" or "spam all the time" mode, but do you mean even more fine-grained control with N + source?

(I just uploaded to latest HEAD + fixed a issue with the spam-button-state, now it saves/restores properly from `localStorage`)

> 2\. Is there a way to improve the stack within console log reporting so it points to the right place of the error? This will allow easier debugging with Dev Tools, but also provide a deep linking from Launcher to IDE.

It's a `console.error` and if you click on it, the stack shows up:

![image](https://github.com/playcanvas/engine/assets/5236548/7cce88bd-ce20-4856-810b-e3f44a30529e)

Then you can click on the file location under `assertType` and you are at the "error source" and can just put a breakpoint:

![image](https://github.com/playcanvas/engine/assets/5236548/b60fa682-4678-470e-9b68-b4f01eeadd13)

So currently it takes two clicks to get to the place of the type error, but I'm not sure what you mean with deep linking from Launcher to IDE, could you elaborate a bit on that?

Thank you again!

Maksims commented:
> Yep, for this there is this button:

It did not work, unfortunately. I tried it, made no difference. Still, had errors spammed:
![image](https://github.com/playcanvas/engine/assets/636263/3a47947b-b299-4338-bca8-90031a41cc9e)

> It's a `console.error` and if you click on it, the stack shows up:

Ah, I did open dev tools *after* the errors were reported, it does not contain stack in such place, and all errors "originated" from the same place.
Is there a way to preserve their original stack? E.g. if the exceptions happen in the user script, but it reports first `playcanvas.rti.js:504`, then it is less clear where the error happened. This is actually an issue already with the current engine, where some errors like providing wrong arguments, are reported from within the engine (that's where they happen), but it is not straight away apparent to the user that it was due to passing wrong arguments. Especially when these arguments are not directly applied, but this is slightly unrelated.
It just highlights an issue: if an error is reported not from the user script, then they first think it is not their doing. So if an error happened somewhere within the user script or engine, the stack should specify errors from there, not from one single place.

Editor can try their best to filter through stack lines, and figure out if it can provide a deep link to a specific script in IDE, as it is really useful.

kungfooman commented:
> It did not work, unfortunately. I tried it, made no difference. Still, had errors spammed:

Hm, maybe the intention of the button isn't clear, when "Spam type reports" is checked it means that every type error is calling `console.error` (and if it is unchecked, the type errors are only logged once). If that doesn't work, it is probably because I updated the file and then it needs a Ctrl+R with DevTools open to clear the cache :thinking:

Thank you for pointing out the stack trace issue when DevTools isn't open from the beginning - I wasn't aware of that, so I will save the stack traces aswell for better inspection (would be nice to see in UI too).

> Editor can try their best to filter through stack lines, and figure out if it can provide a deep link to a specific script in IDE, as it is really useful.

Right, I made a little experiment what kind of information we can get out of the stack trace for Editor integration:

```js
const info = () => {
  try {
    throw new Error(); // Dummy exception to retrieve stack trace
  } catch(e) {
    const {stack} = e;
    if (stack.toLowerCase().includes('script')) {
      const lines = stack.split('\n');
      const i = lines.findIndex(_ => _.includes('ScriptComponent'));
      let scriptLine = lines[i - 1];
      scriptLine = scriptLine.replace(/at/, '').trim().split(' ')[0];
      console.warn('Type error in script: ', scriptLine);
    }
  }
}
info(), false; // false so it doesn't breakpoint here
```

This is "breakpoint code" we can enter here:

![image](https://github.com/playcanvas/engine/assets/5236548/392db817-9e7b-40cf-9467-4d5f25dc7546)

(clicking on the function source)

![image](https://github.com/playcanvas/engine/assets/5236548/92c5917b-2e63-4cfd-8441-aea00d4b6898)

Then you have integrated the "breakpoint code" into the code flow and now you can play around with stack traces:

![image](https://github.com/playcanvas/engine/assets/5236548/7ac533a0-38f1-4033-a0ae-fe62b138b9f0)

Thinking about the editor, we could just add "quick links" to the actual scripts now. Thank you for the suggestion!

kungfooman committed:
Add type reporting mode: never

kungfooman committed:
Merge branch 'main' into rti

# Conflicts:
#	examples/package-lock.json

kungfooman committed:
npm run develop:rti - Support live reload

kungfooman committed:
extras: JSDoc

kungfooman committed:
update RTI version, use index.rti.js for custom types/validations

kungfooman committed:
JSDoc fixes from Splat PR for extras type testing

kungfooman committed:
fix lint

kungfooman committed:
Merge branch 'main' into rti

# Conflicts:
#	rollup.config.mjs

kungfooman committed:
rollup.config.mjs: add rti target

kungfooman committed:
Keep utils/rollup-build-target.mjs without RTI, using rollup-build-target-rti for engine and examples instead (this ensures we only have one function which configures a RTI plugin, so it's mentally easier to navigate)

kungfooman committed:
Fix lint

kungfooman committed:
little jsdoc fix

kungfooman committed:
npm install --save-dev @runtime-type-inspector/plugin-rollup@3.0.0

kungfooman committed:
Fix rollup/acorn error

kungfooman committed:
Merge branch 'main' into rti

# Conflicts:
#	extras/splat/splat-material.js

kungfooman committed:
package-lock.json: npm install (changed jsdoc-typeof-plugin)

kungfooman committed:
npm update @runtime-type-inspector/transpiler (in root and examples/)

kungfooman committed:
npm update @runtime-type-inspector/plugin-rollup

kungfooman committed:
npm update @runtime-type-inspector/plugin-rollup

kungfooman committed:
Merge branch 'main' into rti

# Conflicts:
#	examples/rollup.config.mjs

kungfooman committed:
Update to @runtime-type-inspector/plugin-rollup@3.2.0

kungfooman committed:
Merge branch 'main' into rti

kungfooman committed:
Ignore node_modules for RTI to fix issue with DisplayAnything

kungfooman committed:
Merge branch 'main' into rti

# Conflicts:
#	examples/package.json
#	examples/scripts/generate-standalone-files.mjs
#	rollup.config.mjs

kungfooman committed:
Bump RTI rollup plugin and integrate into examples again for `npm run develop:rti` to work

kungfooman committed:
`cd examples && npm install` to fix `examples/package-lock.json`

kungfooman committed:
Update RTI rollup plugin to 3.2.5

kungfooman committed:
Merge branch 'main' into rti

# Conflicts:
#	examples/package-lock.json
#	examples/package.json
#	examples/rollup.config.mjs
#	examples/scripts/standalone-html.mjs
#	package-lock.json
#	package.json
#	rollup.config.mjs

kungfooman committed:
Reintegrate npm run build:rti

kungfooman committed:
Reintegrate RTI after bunch of merge conflicts

kungfooman committed:
Merge branch 'main' into rti

kungfooman committed:
Fix package.json and package-lock.json from merge conflict

kungfooman committed:
npm i --save-dev @runtime-type-inspector/plugin-rollup@3.2.6

kungfooman committed:
Merge branch 'main' into rti

kungfooman committed:
src/framework/anim/binder/anim-binder.js mini comment fix, you can test this via: `pc.AnimBinder.encode(['spotLight'], 'light', ['color', 'r']);`

kungfooman committed:
Fix custom types

kpal81xd commented:
Im unsure about directly including run time inspector specifics in the engine at best it can support loading a custom engine build but no custom code

kungfooman commented:
Hey @kpal81xd, can you elaborate more on what you mean by adding RTI specifics into the engine? Only a RTI build contains PC specific extra validations (e.g. checking that a `Mat4` contains no `NaN` values)

Thank you for checking this out already!

kungfooman committed:
Merge branch 'main' into rti

# Conflicts:
#	examples/rollup.config.mjs
#	package.json

kungfooman committed:
Bump RTI version

kungfooman committed:
Merge branch 'main' into rti

kungfooman committed:
Bump RTI to 3.3.1

kungfooman committed:
Validate `ArrayBufferView`'s via `ArrayBuffer.isView` as custom type

kungfooman committed:
Bump RTI to 3.3.2

kungfooman committed:
src/index.rti.js: IArguments and ArrayBufferView are checked in RTI itself now

kungfooman committed:
src/index.rti.js: Check for Quaternion floating point precision errors aswell

kungfooman committed:
src/index.rti.js: Add specific warning message to Quat denormalization validation error

kungfooman committed:
Bump RTI to 3.3.3

kungfooman committed:
Fix loaders/gsplat-many example (workers don't do RTI)

kungfooman committed:
Bump RTI to 3.3.7

kungfooman committed:
Merge branch 'main' into rti

kungfooman committed:
Ignore denormalization check for `Quat#mulScalar` aswell

kungfooman commented:
I'm finally coming to a resolution here, it just feels so spammy for many examples still, even tho they run fine (simply based on wrong types in the engine). My main goal is "0 errors" - so developers can start editing examples and only see type errors they introduced themselves during their coding session.

If anyone has time, I would love to have these issues fixed:

 - http://pc.runtimetypeinspector.org/#/graphics/paint-mesh - scopeId seems to be a left over and has to be removed?
 - http://pc.runtimetypeinspector.org/#/user-interface/button-sprite - we are setting `length` of number arrays and move them around and calculating all kinds of stuff with them (leading to vectors filled with `undefined`) - useless calculations, so there is potential for a speedup aswell via small refactor
 - http://pc.runtimetypeinspector.org/#/graphics/lights-baked-a-o - filling of matrices with 'NaN' values
 - significantly reduce error counts, especially the ones that occur every frame

Any volunteers? :sweat_smile:

LeXXik commented:
A good cause 👍 I might tackle some of those next week.

kungfooman committed:
Merge branch 'main' into rti

# Conflicts:
#	package-lock.json

kungfooman committed:
Merge branch 'main' into rti

# Conflicts:
#	package-lock.json
#	package.json

kungfooman committed:
Fix lint for new single quote rule

kungfooman committed:
Merge branch 'main' into rti

# Conflicts:
#	package-lock.json

kungfooman committed:
Sync both engine/ and examples/ package-lock.json

kungfooman committed:
Merge branch 'main' into rti

kungfooman committed:
Small JSDoc fixes

kungfooman committed:
Few more JSDoc fixes

kungfooman committed:
Merge branch 'main' into rti

# Conflicts:
#	examples/rollup.config.mjs
#	examples/utils/plugins/rollup-build-examples.mjs

kungfooman committed:
Merge branch 'main' into rti

# Conflicts:
#	examples/package-lock.json
#	examples/package.json

kungfooman committed:
Fix lint

kungfooman committed:
Merge branch 'main' into rti

kungfooman committed:
Update package-lock.json

kungfooman committed:
Update to RTI v4 (ability to send errors over context boundaries, e.g. error in <iframe> is send to parent window)

kungfooman committed:
Start with Bottombar for bigger debug output integrated in PCUI

kungfooman committed:
Merge branch 'main' into rti

# Conflicts:
#	examples/package-lock.json
#	examples/rollup.config.mjs
#	examples/scripts/build-examples.mjs
#	examples/utils/plugins/rollup-build-examples.mjs
#	package-lock.json
#	package.json

kungfooman committed:
Update and reintegrate RTI and hide Bottombar until finished

kungfooman committed:
fix lint, enable RTI for vercel preview

kungfooman committed:
fix lint round 2

kungfooman commented:
Now the vercel preview comes with RTI :see_no_evil:

willeastcott commented:
> Now the vercel preview comes with RTI 🙈

This is actually really cool - I can imagine going down a rabbit hole fixing a bunch of these issues one weekend! 🚀

kungfooman committed:
Update RTI to 4.0.5 to fix last cross-context cloning issue (e.g. sending data from <iframe> to parent requires it to be clonable)

kungfooman committed:
Fix ShaderVariantParams from spamming

kungfooman committed:
Update canvas and add anim-curve.js (until pr is fixed, I want to remove all type fixes from this PR eventually for minimal changes)

kungfooman committed:
Remove canvas

kungfooman committed:
npm i RTI 4.0.6

kungfooman committed:
Few scene and platform/graphics related type fixes to prevent glTF-related error spam

kungfooman committed:
Merge branch 'main' into rti

# Conflicts:
#	examples/package-lock.json
#	package-lock.json
#	package.json
#	src/scene/shader-lib/programs/standard.js

kungfooman committed:
Fix src/{framework, platform, scene} types

kungfooman committed:
package-lock.json

kungfooman committed:
Fix src/extras/gizmo/view-cube.js group type

kungfooman committed:
Because: options.defines.set('TONEMAP', tonemapNames[options.litOptions.toneMap]);
When options.litOptions.toneMap === -1

kungfooman committed:
Fix two types in /framework/

kungfooman committed:
Merge branch 'main' into rti
